tests/e2e/tests/pilot/testdata/app.yaml.tmpl

```diff
@@ -18,10 +18,10 @@ spec:
   ports:
   - port: 80
     targetPort: {{.port1}}
-    name: http
+    name: foo  # http. This is for testing protocol sniffing.
```

Comment:
@rshriram @crazyxy 

I'm not sure this is the best way to test this. First off, this logic is experimental and is going to be opt-in in pilot.  If we change all our tests over to use unknown ports and run without the flag, I suspect tests will not function properly.

I think we'd be better off for now to have a stand-alone test for this (preferably using the new test framework under `tests/integration/pilot`).  This test would start up its own environment and set the environment variable on pilot (via helm values).

After we switch over to make this the default, we can then decide whether or not we want to migrate other tests to use protocol sniffing.

Sound reasonable?

Reply:
My concern is that adding a single integration test is very difficult to cover existing test cases in pilot e2e test.

Reply:
But what you're doing here is changing all of the tests over to use port sniffing. This is not what we want to do, since this is only an experimental feature.

What tests are we concerned about really? Fundamentally it's reachability for all protocols both with and without MTLS ... does that sound right to you?  If so, then there is a reachability test under `pkg/tests/integration/security/reachability` ... perhaps we should make a copy of it for protocol sniffing?

Reply:
its a bit more than just reachable, as we could have traffic treated as tcp instead of http and still be reachable -- we need to ensure it gets sniffed to the correct protocol. I agree though

Reply:
Yup agreed.  My general point was that it won't be much work to create a new test for this which covers all the areas of concern.

Reply:
I think we also need to cover the test for routing, e.g. the traffic shifting. As previous comment mentioned
> Otherwise, you are only testing a subset of the functionality (reachability and not all the routing issues that we test for).

Reply:
@crazyxy I indicated that we would need to test mTLS, which is part of reachability testing.  Can you explain why we would need to test traffic shifting? How would protocol sniffing impact that?

Reply:
I suggested this change. The existing test has plenty of ports that test all protocols (http/http2/grpc/tcp/etc/) with and without authn, rbac, and a whole host of other things. Throwing the sniffing in the mix is the perfect way to test that the sniffing will work in real life where there is a mixture of ports that we delcared properly and otehrs that are not [i.e.  they will end up using sniffing]. The idea here is to test the interaction of filter chains with sniffing with filter chains that do not use sniffing, to see if there are any adverse side effects - across the board [routing rules, authn/authz, mixer, etc.].

Otherwise, we will never have the confidence to turn this on by default. The end goal is that we will no longer require people to declare protocols in their port names and that everything in Istio continues to work as is. And this is just one way to achieve that. Any separate integration test will not be useful. Duplicating the entire v1alpha3-envoyv2 test suite just for sniffing is also not reasonable as it simply adds yet another test to the long list of test suites we have [not to mention another "required test"].

I do agree that we need to add tests in the new framework but that is orthogonal to this change. We need both.

Reply:
Sure, that makes sense. My concern here is that the change as currently proposed (correct me if I'm wrong) would switch on protocol sniffing across the board. We would no longer be testing the standard path, which would be counter to our goal of risk mitigation.

If we insist on testing everything, then we would have to effectively flag enable this in our tests and duplicate all of our test targets for sniffing.  This would likely be too much of a burden on CI and decrease developer overall mental well being by at least a factor of 2.

So if we don't want to take on that burden, we have to be more selective with which tests we care about. Ultimately, we'll probably want to get to a point where we have one additional test target for sniffing that includes some minimal set of tests.

Reply:
>We would no longer be testing the standard path, which would be counter to our goal of risk mitigation. 

that is not correct. The test enables protocol sniffing only for ports whose protocols cannot be ascertained through their port names. So, in the app.yaml.tmpl, only foo will have protocol sniffing, while the rest will not use the protocol sniffing logic at all [they have ports explicitly named as http-*, tcp, tcp-*].

Reply:
The new `app.yaml.tmpl` changes the 2 http ports to unknown. I'm assuming that the majority of our tests would be using these ports, since we're typically sending http traffic. So each test that uses HTTP would now only test the sniffing path. Additionally we wouldn't be testing other protocols, such as grpc.

Reply:
yes, I told @crazyxy to change just one port to foo and not both. Also, 
>Additionally we wouldn't be testing other protocols, such as grpc.
is not correct. WE have other ports named http2, grpc etc. and all of those are fully tested. And to add to this, we have dedicated tests for grpc (grpc_test.go) which  are passing with or without sniffing on the ports marked as grpc/http2